### PR TITLE
Increase resilience to spoofed Version Negotiation packets

### LINKED ISSUE #523: Version Negotiation validation
@janaiyengar observes,

> We haven't nailed down rules for when an endpoint receives a VN with a packet number that does not match any sent packets.

This is true.  We should write that down.

### PR DESCRIPTION
**Pull request by martinthomson**
There's two layers of defense here:

1. Version Negotiation echoes 31-bits of entropy (maybe more if you consider the version to be at all unpredictable, or the client's source address to contain any entropy).

2. If validation fails and version negotiation happened, encourage the client to try again ignoring any Version Negotiation packets that match the failed profile.

Closes #523.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -1228,6 +1235,18 @@ client MUST terminate with a QUIC_VERSION_NEGOTIATION_MISMATCH error code if
...
 
+If the client receives a Version Negotiation packet and these validation checks
+subsequently fail, it is likely that the client received a spoofed Version
+Negotiation packet.  A client MAY attempt to create a new connection and ignore
+any Version Negotiation packets that match those that caused the connection to
+fail.
+
+Note:
+
+: The client cannot rely on the version list from the transport parameters.
```
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: I'd rephrase this -- the client certainly *can* rely on that list -- once the handshake has completed.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -931,13 +931,20 @@ If the packet contains a version that is acceptable to the server, the server
...
+
+A client that receives a valid Version Negotiation packet selects an acceptable
+protocol version from those listed by the server.  The client then reattempts to
+create a connection using that version.  Though the contents of a packet might
+not change in response to version negotiation, a client MUST choose a new packet
+number it uses on every packet it sends.
+
+If the server does not list an acceptable version, the client MAY ignore the
+Version Negotiation packet.  This might reduce the likelihood that a spoofed
+Version Negotiation packet can be used to disrupt connection establishment.
```
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: This seems very painful if there is a real mismatch.  We've already blocked out injected packets, and I suspect that legitimate failures will be more common.

Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: +1. I would expect the common case to be mismatch, since an attacker can easily generate a VN packet that supports a large number of known versions (minus the client's chosen).

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: The client has to rely on the list; it just cannot validate the list until the end of the handshake.

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -1228,6 +1235,18 @@ client MUST terminate with a QUIC_VERSION_NEGOTIATION_MISMATCH error code if
 version negotiation occurred but it would have selected a different version
 based on the value of the supported_versions list.
 
+If the client receives a Version Negotiation packet and these validation checks
+subsequently fail, it is likely that the client received a spoofed Version
+Negotiation packet.  A client MAY attempt to create a new connection and ignore
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I don't think we know this to be likely a result of spoofing, so I wouldn't say it yet.

File: draft-ietf-quic-transport.md (hunk 2 above)
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: I'll cut this paragraph.

File: draft-ietf-quic-transport.md (hunk 3 above)
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I completely agree that something is broken in this case, but the text doesn't say that. It says "it is likely that the client received a spoofed Version Negotiation packet."

**Comment by martinthomson** [comment]
I ended up dropping this for some reason, but I've updated it now.

**Comment by mikkelfj** [comment]
What makes version negotiation special compared to all other first packets from a server?

**Comment by martinthomson** [comment]
OBE: we decided not to worry about spoofed version negotiation too much.  #724 covers the bit we want to keep here.
